### DISCUSSION TITLE
Next/image with Cloudinary

### DISCUSSION THREAD

**Discussion by josh-unwin** [vercel/next.js#18374] [Help]
Hi everyone,

Just diving into the new goodies in Next 10, looks fantastic!

I'm wondering if there is much benefit in using next/image when hosting images via Cloudinary? At the moment Cloudinary is generating different versions of images when uploaded via my CMS, but I'd be interested in changing that to just generating a single image and then having Next generate other versions on the fly after to see what improvements there are.

I tried to have a quick play with it earlier and am getting the error:
```
Invalid src prop on `next/image`, hostname is not configured under images in your `next.config.js`
```

My next.config.js:
```
module.exports = {
  webpack: (config) => {
      config.module.rules.push({
          test: /\.md$/,
          use: 'raw-loader'
      });
      return config
  },

  images: {
    deviceSizes: [320, 420, 768, 1024, 1200],
    iconSizes: [],
    domains: ["https://res.cloudinary.com/"],
    path: '/_next/image',
    loader: 'default',
  },
}
```

**Comment by jamesmosier** [comment]
Hi @josh-unwin. Have you taken a look at the [Loader](https://nextjs.org/docs/basic-features/image-optimization#loader) option which supports Cloudinary? 

Based on that section, it seems like you can still use the `<Image>` component but with relative urls to your image in Cloudinary. That would then circumvent `next/image` from doing the image processing.

**Reply by josh-unwin** [reply]
I was wondering more if I can actually use the next/image image processing but pull the source image from Cloudinary, perhaps that is a strange thing to do, but if not then I'm not sure if there's any benefit using this new `next/image` versus my current setup which is using `cloudinary-react`.

**Reply by ProjectINT** [reply]
cloudinary-react it's 45kb in bundle size

**Comment by enochndika** [comment]


**Reply by josh-unwin** [reply]
Thanks for that, I haven't quite got it working yet (url isn't coming out right) but probably just a case of playing around.
I was actually more interested in whether there's any point in using `next/image` as the loader to take care of the processing, if not I'm not sure what benefit there could be over using `cloudinary-react` which is my current setup.

**Comment by uinstinct** [comment]
I noticed this solution was working
```js
module.exports = {
    images: {
        domains: ["res.cloudinary.com"],
    },
}
```
Adding the `loader` as _cloudinary_ prepends `_next/static` folder as `localhost`.

**Reply by josh-unwin** [reply]
If I'm understanding correct (which I may well not be :P), this is the answer to my question. I believe through this I'm able to pull up the source image from Cloudinary but do the optimisation through next/image.

**Comment by irpic** [comment]
If you want to use it with a CMS for example, and your images are located on localhost, add localhost to domains on images -> next.config.js file:
 
`images: {
    domains: ['localhost']
}`

And use the absolut path for src on Image component like this:

`<Image src='http://localhost:1000/uploads/img.jpg' width={500} height={500} alt="Picture of the author" />`

**Reply by uinstinct** [reply]
Thanks. Yeah that is now working!

Are there any methods to style these **images**?

**Reply by irpic** [reply]
What do you want to style exactly? You can use css via className for example, on the Image component.

You can set props on the Image component as well.

[Here are the available props](https://nextjs.org/docs/api-reference/next/image)

**Reply by uinstinct** [reply]
I want to style the **`img`** itself.

I cannot see any _props_ for `className` or inline `styles` to the *`next/image`* component.

**Comment by erichodges** [comment]
I would prefer to have Cloudinary do the transformations.

I setup the next.config.js:
```
module.exports = {
  images: {
    loader: 'cloudinary',
    // path: 'cloudinary://774874966656811:4q-UDjejk_QY8Yn7DPUa0dypTow@dn3baky3u'
    domains: ['res.cloudinary.com/']
  }
};
```
A sample image from my account is here:
https://res.cloudinary.com/dn3baky3u/image/upload/v1580246844/gatsby-cloudinary/evora-gt-side.jpg

I also tried putting the above path in the Image component src.
What am I doing wrong?

**Reply by dackers86** [reply]
I'm not sure if this would cause any performance issue, still trying to get my PWA to 100 on PageSpeed!

Using the next loader you can simply add...

```js
 images: {
      domains: ["res.cloudinary.com"],
    },
```

Also use a wrapper component for now....

```js
import Image from "next/image";

const ImageComponent = ({
  src,
  width = 240,
  height = 750,
  loading = "lazy",
}) => {
  const hostUrl = `https://res.cloudinary.com/${cloundinary_id}/image/upload/q_100,w_${width},h_${height}`;

  return (
    <Image
      class="object-center object-cover w-full h-full"
      src={`${hostUrl}/${src}`}
      alt="Picture of the author"
      width={width}
      height={height}
      layout="responsive"
      quality={10}
      loading={loading}
    />
  );
};

export default ImageComponent;
```

**Reply by erichodges** [reply]
Thank you. I fixed the config file.

I tried a couple of things, neither of which worked. Getting 404 for the image.

```
import Image from 'next/image';
import React from 'react';

const ImageComponent = ({
  src,
  width = 240,
  height = 750,
  loading = 'lazy'
}) => {
  const hostUrl = `https://res.cloudinary.com/dn3baky3u/image/upload/v1580246844/gatsby-cloudinary/evora-gt-side`;
  // https://res.cloudinary.com/dn3baky3u/image/upload/v1580246844/gatsby-cloudinary/evora-gt-side.jpg
  // https://res.cloudinary.com/dn3baky3u/image/upload/v1580246844/gatsby-cloudinary/evora-gt-side
  return (
    <Image
      className='object-center object-cover w-full h-full'
      src={`${hostUrl}/${src}`}
      alt='Picture of the author'
      width={width}
      height={height}
      layout='responsive'
      quality={10}
      loading={loading}
    />
  );
};

export default function Home() {
  return (
    <div>
      <ImageComponent />
    </div>
  );
}
```
Note that I used my cloudinary cloud name where you used cloundinary_id. Not sure if that is what you meant.
I want to have cloudinary return an optimized image for a responsive page and not fix the width/height.

**Reply by dackers86** [reply]
@erichodges You have included the image `src` in your host url, please try.

const hostUrl = `https://res.cloudinary.com/dn3baky3u/image/upload/v1580246844/gatsby-cloudinary`;

And pass in `evora-gt-side` as the src as a prop

**Reply by dackers86** [reply]
(Optionally) Using width and height through Cloudinary will also match Next when resizing, this should vastly improve page performance when loading the image. Your max values should be passed through to allow Next to produce valid srcsets.

**Reply by erichodges** [reply]
Ah, that makes sense and now I get an image. 

I'm wondering, though, about `layout=responsive` etc. Instead of doing width/height, I `used layout = 'fill'` and then `layout={layout}` in the Image component. 

I will read up on using q_auto and f_auto in the URL. At the moment, I only get the full sized image.

**Reply by erichodges** [reply]
Ah ok, max values are doable, I will try that.

**Reply by erichodges** [reply]
I set it up this way:
```jsx
const ImageComponent = ({
  src = 'evora-gt-side',
  width = 2400,
  height = 1200,
  loading = 'lazy'
}) => {
  const hostUrl = `https://res.cloudinary.com/dn3baky3u/image/upload/w_${width},h_${height},f_auto,dpr_auto/v1580246844/gatsby-cloudinary`;
  // https://res.cloudinary.com/dn3baky3u/image/upload/v1580246844/gatsby-cloudinary/evora-gt-side.jpg
  // https://res.cloudinary.com/dn3baky3u/image/upload/v1580246844/gatsby-cloudinary/evora-gt-side
  return (
    <Image
      className='object-center object-cover w-full h-full'
      src={`${hostUrl}/${src}`}
      alt='Picture of the author'
      layout='responsive'
      quality={10}
      loading={loading}
      width={width}
      height={height}
    />
  );
};
```
The smallest image I could get was 1060px wide. Is it possible to get anything smaller? When I look at the srcset, I see that there is one at 640px.
Update: I looked again, and I'm only getting that 1060px image. Maybe I made a mistake somewhere?

**Reply by DoctorDerek** [reply]
@erichodges I know your post is about 6 months old, but I wanted to mention that you can control the device sizes https://nextjs.org/docs/basic-features/image-optimization#device-sizes

**Reply by colbyfayock** [reply]
hey @erichodges can you DM me on twitter @colbyfayock or email me at hello@colbyfayock.com? have a question for you 🙏  TIA

**Comment by uinstinct** [comment]
I really think that there should be this feature to use _next/image_ during `next export`, so that we could statically serve it.

**Comment by rebeccapeltz** [comment]
I've created a code sandbox to load a cloudinary image https://codesandbox.io/s/lingering-fog-7irj6?file=/pages/index.js. I'm getting an error that my domain name is not included but I have a next.config.js in the sandbox that lists cloudinary as the domain

```js
module.exports = {
  images: {
    domains: ["res.cloudinary.com"]
  }
};
```

![image](https://user-images.githubusercontent.com/354444/107593232-be463900-6bc3-11eb-8ae1-9a96314371ac.png)

**Reply by erichodges** [reply]
@rebeccapeltz  I forked your sandbox an got it to work by just putting in the URL in `src='<URL-here>'`

https://codesandbox.io/s/great-glitter-01gj9?file=/pages/index.js

Maybe that will work for you?

**Comment by sudip1072** [comment]
Follow below steps to resolve the issue. 

1.setup cloudinary domain into next.config.js in below format.
images: {
domains: ['res.cloudinary.com'],
},

2.back to frontend , define width and height and layout in your next/image component.

**Comment by Utzel-Butzel** [comment]
Thanks for the help! I had a bit trouble to get the width & height server side, because it looks like everyone else is using static dimensions. For getting the right dimensions server side I now check the [Cloudinary API](https://cloudinary.com/documentation/transformation_reference) using  the`get_info` flag.


Example using `cloudinary-core` to generate the url.

```jsx

import { Cloudinary } from "cloudinary-core";

async function probe(src) {
  // Use function to get cloud_name and name from cloudinary image url
  const { cloudName, name, isVideo } = urlGenerator(src);
  var cl = new Cloudinary({ cloud_name: cloudName, secure: false });

  const url = cl.url(name, {
    protocol: "https:",
    width: 300,
    flags: ["getinfo"],
  });

  const res = await fetch(url, {
    method: "GET",
  });

  const result = await res.json();

  return { width: result.input.width, height: result.input.height };
}
```

I'm using the code in a [rehype](https://mdxjs.com/docs/extending-mdx/) plugin for mdx where I don't know the image size beforehand but need it for the aspect ratio in my [custom lightbox](https://anabox-smart.de/posts/components).

This is the code I use in the rehype plugin: https://gist.github.com/Utzel-Butzel/d70df2a60b2de503fb9c9aa226cb73ba

**Comment by colbyfayock** [comment]
I know this is an older question but people still ask this a lot, so here are my thoughts.

the Next Image component provides some cool capabilities that have crossover with Cloudinary, particularly the optimization part and serving modern formats. Rather than trying to compare the two for that, probably worth noting what Next Image _can't_ do that Cloudinary can, like more advanced transformations, media management, analysis, AI-based cropping, etc etc 

That said, I think most people asking this question are just wondering about the optimization side where if you're using Cloudinary URLs on top of Next.js processing you're really doing double duty, which is _fine_ but not really going to yield you anything if you're doing the same thing twice.

If you want to use Next Image, you have really 2 options* (a 3rd being a custom loader), here are examples:

### Built-in Loader

This uses Cloudinary tech to handle things like transformations and the optimization.

For instance with the basic settings I have the resulting URL of one of my images on the demo below would be:

```
https://res.cloudinary.com/colbycloud-examples/f_auto,c_limit,w_3840,q_auto/examples/beach_udmhbt.jpg
```

* Demo: https://cloudinary-nextjs-next-image-loader.netlify.app/
* Code: https://github.com/colbyfayock/cloudinary-examples/tree/main/examples/nextjs-next-image-loader
* Relevant Next.js code: https://github.com/vercel/next.js/blob/160da6b80692738c575c3401c6abbfc5f52f721f/packages/next/client/image.tsx#L997

```
const nextConfig = {
  reactStrictMode: true,
  images: {
    loader: 'cloudinary',
    path: `https://res.cloudinary.com/<Your Cloud Name>`
  }
}
```

### Adding it as a Domain

This utilizes the full Next Image processing it just allows you to use an image from Cloudinary as the source. The double duty I alluded to before.

For instance with the basic settings I have the resulting URL of one of my images on the demo below would be:

```
/_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fcolbycloud-examples%2Fimage%2Fupload%2Fv1654534626%2Fexamples%2Fbeach_udmhbt.jpg&w=3840&q=75
```

* Demo: https://cloudinary-nextjs-next-image-domain.netlify.app/
* Code: https://github.com/colbyfayock/cloudinary-examples/tree/main/examples/nextjs-next-image-domain

```
const nextConfig = {
  reactStrictMode: true,
  images: {
    domains: ['res.cloudinary.com']
  }
}
```

### Custom Loader

You can supply your own loader to the Next Image component to go beyond what the default does. I've been considering creating an "official yet unofficial" Cloudinary loader to help people with this

### Other Options

As you've noticed Cloudinary also has it's own [React SDK](https://cloudinary.com/documentation/react_integration) that gives some more advanced capabilities as well.

### From a Headless CMS

There are a few different integration points where you can use Cloudinary for future reference. I'd be curious about what solution you're using where the CMS is generating a few images. You can create new source images with trasnformations, but a big benefit of Cloudinary is indeed the ability to do on-the-fly transformations, so for instance if from your headless CMS you have 1 image at a high resolution, you can add all of the URL params you need to that Cloudinary URL for whatever you need in your application, such as resizing, cropping, or whatever making it pretty easy to have that flexibilty in your app. The resulting images would then be cached and served from the CDN.

---

Happy to try to answer any other questions or add other examples to my [cloudinary-examples repo](https://github.com/colbyfayock/cloudinary-examples) if there are other use cases I can help people with!

**Reply by colbyfayock** [reply]
If anyone's interested, I took this a step further and created an experimental wrapper component!

https://github.com/colbyfayock/next-cloudinary

Would love feedback, ideas, and suggestions 🙌

**Comment by rikers** [comment]
you made things clear @colbyfayock thank you (Love your YT chanel btw!). 
Here is my take on the question, starting with a bit of context.

1) On the CMS side, I have the ability to add and edit products. On a given product, I can crop my image:

![image](https://user-images.githubusercontent.com/15691035/186138252-5d197e9d-89a3-453d-8b5e-a3dbc4eb0462.png)

2) The data is exposed on a REST API endpoint such as https://app-domain/api/products
Here's a sample of the resulting data:

```json
{
        "id": 29920,
        "type": "image",
        "title": "my title",
        "description": "my description",
        "width": 1797,
        "height": 1198,
        "public_id": "a_folder/cq7ptxe2btyqngd89pfw",
        "original_url": "https://res.cloudinary.com/my_cloud_name/a_folder/cq7ptxe2btyqngd89pfw.jpg",
        "crop_x": 605,
        "crop_y": 0,
        "crop_w": 1031,
        "crop_h": 1198
}
```

3) On my page, products are displayed in a grid, given the following code:

```jsx
function HomePage(props) {
    // ...

    return (
        
        // ...

        <GridFourCols
            CustomItemsContainerTag='div'
            headerSlot={<h3>The header slot</h3>}
            footerSlot={<h3>The footer Slot</h3>}
        >
            {props.products.map((product) => {
                return (
                    <Cards.Product
                        CustomTagContainer='div'
                        key={product.id}
                        img={{
                            alt: '',
                            objectFit: 'cover',
                            layout: 'responsive',
                            sizes: PRODUCT_SIZES,
                            src: product.public_id,
                            // using closure, customCloudinaryLoader returns the actual custom loader function
                            loader: customCloudinaryLoader(product),
                            width: 260,
                            height: 302,
                        }}
                        // ... Other props
                    />
                );
            })}
        </GridFourCols>

        // ...

    );
}
```

4) here is the code for my custom loader: 

```js
import buildCloudinaryUrl from '@/utils/buildCloudinaryUrl';

const customCloudinaryLoader = ({
    crop_x,
    crop_y,
    crop_w,
    crop_h
}) => {
    return ({
        src,
        width: deviceSize
    }) => {
        const transformations = {
            x: crop_x,
            y: crop_y,
            w: crop_w,
            h: crop_h,
        };

        return `https://res.cloudinary.com/${process.env.NEXT_PUBLIC_CLOUDINARY_CLOUD_NAME}${buildCloudinaryUrl(src, deviceSize, transformations)}`;
    };
};

export default customCloudinaryLoader;
```

5) the custom loader uses this "little" helper function:

```js
import { isNumber } from 'lodash';

const validTransformations = [
    'x',
    'y',
    'w',
    'h',
];

/**
 * It takes a public_id and an optional transformations object and returns a URL to the image on Cloudinary
 * @param {string} public_id - The public_id of the image you want to transform.
 * @param {number} deviceSize - See: https://nextjs.org/docs/api-reference/next/image#device-sizes
 * @param {object} transformations - An object containing valid transformations.
 * @param {number} transformations.crop_x - 0 or a positive integer.
 * @param {number} transformations.crop_y - 0 or a positive integer.
 * @param {number} transformations.crop_w - A positive integer.
 * @param {number} transformations.crop_h - A positive integer.
 * @returns A URL string.
 * For this project, we're using Cloudinary as a basic CMS with it's URL API possibilities.
*/
const buildCloudinaryUrl = (
    public_id,
    deviceSize,
    transformations = {}
) => {
    const CROPPED_IMAGE_WIDTH  = transformations.x + transformations.w;

    let transformationsString = Object.keys(transformations)
        .map(key => {
            if (validTransformations.findIndex(validTransformation => validTransformation === key) === -1) {
                console.error(`❌ ~ Transformation ${key} is not valid. Valid Transformations: `, validTransformations);
                return null;
            }

            const val = transformations[key];

            switch(true) {
                case key === 'w' && val === 0:
                case key === 'h' && val === 0:
                    console.error(`❌ ~ "w" and "h" transformations must be a positive integer.`);
                    return null;

                case !isNumber(val):
                case val === null:
                case val === undefined:
                case val < 0:
                    console.error(`❌ ~ Transformation ${key} must be a positive integer. you provided: ${val}. Type: ${typeof val}.`);
                    return null;

                default:
                    return `${key}_${transformations[key]}`;
            }
        })
        .filter((entry => entry !== null))
        .join(',');

    // Append c_crop at the beginning of the string
    transformationsString = `f_auto,q_auto,c_crop,${transformationsString}`;

    // Chained transformations (reduce image width and height by preserving the ratio)
    // using the deviceSize passed to the loader function to generate the srcset attribute
    let chainedTransformationString = '';

    if (CROPPED_IMAGE_WIDTH >= deviceSize) {
        chainedTransformationString = `c_scale,w_${deviceSize}`;
    } else {
        chainedTransformationString = `c_scale,w_${CROPPED_IMAGE_WIDTH}`;
    }

    return `/${transformationsString}/${chainedTransformationString}/${public_id}`;
};

export default buildCloudinaryUrl;
```

6) And here is the resulting html:
```html
<img
    alt=""
    src="https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_78,y_0,w_1031,h_1198/c_scale,w_1109/a_folder/cq7ptxe2btyqngd89pfw"
    decoding="async"
    data-nimg="responsive"
    style="
        position: absolute;
        top: 0;
        left: 0;
        bottom: 0;
        right: 0;
        box-sizing: border-box;
        padding: 0;
        border: none;
        margin: auto;
        display: block;
        width: 0;
        height: 0;
        min-width: 100%;
        max-width: 100%;
        min-height: 100%;
        max-height: 100%;
        object-fit: cover;
    "
    sizes="(min-width: 1160px) 260px, (min-width: 768px) calc(33vw - 78px), (min-width: 414px) calc(50vw - 54px), 100vw"
    srcset="
        https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_78,y_0,w_1031,h_1198/c_scale,w_640/a_folder/cq7ptxe2btyqngd89pfw   640w,
        https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_78,y_0,w_1031,h_1198/c_scale,w_750/a_folder/cq7ptxe2btyqngd89pfw   750w,
        https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_78,y_0,w_1031,h_1198/c_scale,w_828/a_folder/cq7ptxe2btyqngd89pfw   828w,
        https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_78,y_0,w_1031,h_1198/c_scale,w_1080/a_folder/cq7ptxe2btyqngd89pfw 1080w,
        https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_78,y_0,w_1031,h_1198/c_scale,w_1109/a_folder/cq7ptxe2btyqngd89pfw 1200w,
        https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_78,y_0,w_1031,h_1198/c_scale,w_1109/a_folder/cq7ptxe2btyqngd89pfw 1920w,
        https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_78,y_0,w_1031,h_1198/c_scale,w_1109/a_folder/cq7ptxe2btyqngd89pfw 2048w,
        https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_78,y_0,w_1031,h_1198/c_scale,w_1109/a_folder/cq7ptxe2btyqngd89pfw 3840w
    "
/>
```

**Feedbacks will be appreciated, as it's a quick and first try on the subject!**

**Reply by colbyfayock** [reply]
i think you have a nice-looking solution there, what you're doing make sense

i'd love to make this easier for you and others with a similar use case

i started an experimental Next.js Cloudinary component that wraps the existing Image and extends with Cloudinary tech

it's still early and i don't have that cropping capability you're using yet (ill add that, good example), but would love thoughts if you have time to poke around it

https://github.com/colbyfayock/next-cloudinary

the README is sparse at the moment but you can find examples on the demo site:

https://github.com/colbyfayock/next-cloudinary/blob/main/example/pages/index.js
https://next-cloudinary-example.vercel.app/

**Reply by rikers** [reply]
Thank you for the reply. I'll sure take a look on it!

I decided to take it a step further by re-using the logic in cases where a picture element is needed (art direction driven image, with different sources for example).

I guess the component is a bit dirty for now, but it works...

```jsx
import {
    useCallback,
    useMemo,
    useState
} from 'react';

import DEVICE_SIZES from '@/constants/device-sizes';
import customCloudinaryLoader from '@/utils/customCloudinaryLoader';
import styles from './Picture.module.scss';

const Picture = ({
    onLoad,
    onError,
    extraClasses = [],
    alt = '',
    objectFit='cover', // or 'fill', 'contain', 'none', 'scale-down'
    loading='lazy', // or 'eager'
    images,
    children
}) => {
    const [
        isReady,
        setIsReady
    ] = useState(false);

    const onLoadCallback = useCallback((e) => {
        setIsReady(true);
        typeof onLoad === 'function' && onLoad(e);
    }, [onLoad]);

    const onErrorCallback = useCallback((e) => {
        setIsReady(true);
        typeof onError === 'function' && onError(e);
    }, [onError]);

    const ref = useCallback((node) => {
        node?.complete && setIsReady(true);
    }, []);

    const containerClasses = useMemo(() => [
        styles.container,
        ...(extraClasses.length ? extraClasses : []),
        ...(isReady ? [styles.ready] : [])
    ].join(' '), [
        isReady,
        extraClasses
    ]);

    const imgClasses = useMemo(() => [
        styles.img,
        styles[objectFit]
    ].join(' '), [objectFit]);

    const defaultImg = images.find(img => !img.media);

    const defaultImgLoader = customCloudinaryLoader({
        crop_x: defaultImg.crop_x,
        crop_y: defaultImg.crop_y,
        crop_w: defaultImg.crop_w,
        crop_h: defaultImg.crop_h,
    });

    let defaultImgSrc = '';

    let defaultImgSrcSet = DEVICE_SIZES
        .map((ds, index) => {
            let string = `${defaultImgLoader({
                src: defaultImg.public_id,
                width: ds
            })} ${ds}w`;

            if (index === 0) {
                defaultImgSrc = string;
            }

            return string;
        })
        .join(', ');

    return (
        <div className={containerClasses}>
            <picture className={styles.picture}>
                {
                    images.map((img) => {
                        if (!img.media) return;

                        let loader = customCloudinaryLoader({
                            crop_x: img.crop_x,
                            crop_y: img.crop_y,
                            crop_w: img.crop_w,
                            crop_h: img.crop_h,
                        });

                        let srcSet = DEVICE_SIZES
                            .map(ds => `${loader({
                                src: img.public_id,
                                width: ds
                            })} ${ds}w`)
                            .join(', ');

                        return (
                            <source
                                key={img.id}
                                media={img.media}
                                srcSet={srcSet}
                                sizes={img.sizes}
                            />
                        );
                    })
                }

                <img
                    ref={ref}
                    onLoad={onLoadCallback}
                    onError={onErrorCallback}
                    src={defaultImgSrc}
                    srcSet={defaultImgSrcSet}
                    sizes={defaultImg.sizes}
                    alt={alt}
                    className={imgClasses}
                    loading={loading}
                />
            </picture>
            {children}
        </div>
    );
};

export { Picture };
```

Usage: 

```jsx
<Picture
    alt='my picture alt text'
    objectFit='cover'
    loading='eager'
    images={
        [
            // image_cover
            {
                ...image_cover,
                sizes: '100vw',
                media: `(min-width: ${BREAKPOINTS.md}px)`,
                useCloudinary: true,
            },
            // image_cover_mobile
            {
                ...image_cover_mobile,
                sizes: '100vw',
                useCloudinary: true,
            }
        ]
    }
/>
```

useCloudinary does nothing for now but I might tweak the component so other img sources can be used, without processing the infos trough the custom loader.

datas look like:

```json
"image_cover": {
    "id": 29992,
    "public_id": "a_folder/dw848ablneglp3swqor2",
    "original_url": "https://res.cloudinary.com/my_cloud_name/a_folder/dw848ablneglp3swqor2.png",
    "crop_x": 362,
    "crop_y": 354,
    "crop_w": 1737,
    "crop_h": 1095
},
"image_cover_mobile": {
    "id": 29595,
    "public_id": "a_folder/jkkipyagxxsbia8fvkj4",
    "original_url": "https://res.cloudinary.com/my_cloud_name/a_folder/jkkipyagxxsbia8fvkj4.jpg",
    "crop_x": 1514,
    "crop_y": 0,
    "crop_w": 785,
    "crop_h": 1699
}
```

And here is the resulting HTML:

```html
<div class="Picture_container__EuxK3 Picture_ready__sisGF">
    <picture class="Picture_picture__Ovb9p">
        <source
            media="(min-width: 1024px)"
            srcset="https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_362,y_354,w_1737,h_1095/c_scale,w_360/a_folder/dw848ablneglp3swqor2 360w,
                    https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_362,y_354,w_1737,h_1095/c_scale,w_640/a_folder/dw848ablneglp3swqor2 640w,
                    https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_362,y_354,w_1737,h_1095/c_scale,w_750/a_folder/dw848ablneglp3swqor2 750w,
                    https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_362,y_354,w_1737,h_1095/c_scale,w_828/a_folder/dw848ablneglp3swqor2 828w
                    https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_362,y_354,w_1737,h_1095/c_scale,w_1080/a_folder/dw848ablneglp3swqor2 1080w,
                    https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_362,y_354,w_1737,h_1095/c_scale,w_1200/a_folder/dw848ablneglp3swqor2 1200w,
                    https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_362,y_354,w_1737,h_1095/c_scale,w_1920/a_folder/dw848ablneglp3swqor2 1920w,
                    https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_362,y_354,w_1737,h_1095/c_scale,w_2048/a_folder/dw848ablneglp3swqor2 2048w,
                    https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_362,y_354,w_1737,h_1095/c_scale,w_2099/a_folder/dw848ablneglp3swqor2 3840w"
            sizes="100vw"
        >
        
        <img
            src="https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_360/a_folder/jkkipyagxxsbia8fvkj4 360w"
            srcset="https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_360/a_folder/jkkipyagxxsbia8fvkj4 360w,
            https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_640/a_folder/jkkipyagxxsbia8fvkj4 640w,
            https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_750/a_folder/jkkipyagxxsbia8fvkj4 750w,
            https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_828/a_folder/jkkipyagxxsbia8fvkj4 828w,
            https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_1080/a_folder/jkkipyagxxsbia8fvkj4 1080w,
            https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_1200/a_folder/jkkipyagxxsbia8fvkj4 1200w,
            https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_1920/a_folder/jkkipyagxxsbia8fvkj4 1920w,
            https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_2048/a_folder/jkkipyagxxsbia8fvkj4 2048w,
            https://res.cloudinary.com/my_cloud_name/f_auto,q_auto,c_crop,x_1514,y_0,w_785,h_1699/c_scale,w_2299/a_folder/jkkipyagxxsbia8fvkj4 3840w"
            sizes="100vw"
            alt="my picture alt text"
            class="Picture_img__Srt_C Picture_cover__4_bzl"
            loading="eager"
        >
    </picture>
</div>
```

**Hope this will help !**
